### PR TITLE
Fixes #28720 - connect to mongo for content migrations

### PR DESCRIPTION
**Pull request by wbclark**
https://github.com/theforeman/puppet-pulpcore/issues/33 could also be addressed by expanding on this but would introduce additionally parameters that don't yet seem necessary.

### CODE REVIEW

**Comment by wbclark** [comment]
> IMHO these should be parameters since mongodb can be external. Not 100% sure we support that in the katello scenario, but at that level we have all the correct information. Also if you don't pass them in by default, you don't give the impression that it's configured.
> 
> The way I would design this is to create `pulpcore::plugin::migration`. That can drop in the required configuration and ensure the migration package is installed. If you want a migration to be set up, you include the plugin class with the right parameters and it's all set up correctly.

https://github.com/theforeman/puppet-pulpcore/blob/master/manifests/plugin.pp might need to be reworked a bit because it currently always tries to install a package by the name of `python3-pulp-${title}` but I think that is doable.

**Comment by ehelms** [comment]
We do indeed support external MongoDB, so we would need this to support:

 1) connecting to an external MongoDB
 2) connecting to MongoDB with some authentication

File: manifests/config.pp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
   concat::fragment { 'base':
     target  => 'pulpcore settings',
+#    content => template('pulpcore/settings.py.erb', 'pulpcore/settings-mongodb.py.erb'),
```
Comment (ekohl) [inline manifests/config.pp]: Not a fan of commented code :)

File: templates/settings-mongodb.py.erb [hunk 2]
```diff
@@ -0,0 +1 @@
+PULP2_MONGODB = {"name": "<%= scope['pulpcore::mongo_db_name'] %>", "seeds": "<%= scope['pulpcore::mongo_db_seeds'] %>", "ssl": <%= scope['pulpcore::mongo_db_ssl'] %>}
```
Comment (ekohl) [inline templates/settings-mongodb.py.erb]: I'd suggest so format this in multiple lines.

File: manifests/init.pp [hunk 3]
```diff
@@ -66,6 +66,9 @@
   String $postgresql_db_password = extlib::cache_data('pulpcore_cache_data', 'db_password', extlib::random_password(32)),
   String $django_secret_key = extlib::cache_data('pulpcore_cache_data', 'secret_key', extlib::random_password(32)),
   Integer[0] $redis_db = 8,
+  String $mongo_db_name = 'pulp_database',
```
Comment (ekohl) [inline manifests/init.pp]: These should be documented as `@param`

Comment (wbclark) [reply manifests/init.pp]: Will do, thanks

File: manifests/config.pp (hunk 1 above)
Comment (wbclark) [reply manifests/config.pp]: Just a temporary test. I can't run them locally due to some environmental issues that I'm putting off resolving until after this is complete.

File: templates/settings-mongodb.py.erb (hunk 2 above)
Comment (wbclark) [reply templates/settings-mongodb.py.erb]: Will do. I'm planning as well to put this back in the main `settings.py.erb` template with a conditional.

**Comment by ekohl** [comment]
https://github.com/theforeman/foreman-packaging/pull/4660 should provide the plugin package.

**Comment by wbclark** [comment]
Will look at moving https://github.com/theforeman/puppet-pulpcore/pull/57/files#diff-9126f4197436df08d4e311a62a5daf47R22-R26 into manifests/plugin/migration.pp

File: spec/classes/plugin_file_spec.rb [hunk 4]
```diff
@@ -15,6 +15,7 @@
           is_expected.to compile.with_all_deps
           is_expected.to contain_pulpcore__plugin('file')
             .that_subscribes_to('Class[Pulpcore::Install]')
+          is_expected.to contain_pulpcore__plugin('file')
```
Comment (ekohl) [inline spec/classes/plugin_file_spec.rb]: This shouldn't be needed

File: spec/classes/plugin_container_spec.rb [hunk 5]
```diff
@@ -15,6 +15,7 @@
           is_expected.to compile.with_all_deps
           is_expected.to contain_pulpcore__plugin('container')
             .that_subscribes_to('Class[Pulpcore::Install]')
+          is_expected.to contain_pulpcore__plugin('container')
```
Comment (ekohl) [inline spec/classes/plugin_container_spec.rb]: This shouldn't be needed

File: spec/classes/plugin_migration_spec.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      it { is_expected.to contain_pulpcore__plugin('migration') }
+
+      context 'with pulpcore' do
+        let(:pre_condition) { 'include pulpcore' }
+
+        it do
+          is_expected.to compile.with_all_deps
+          is_expected.to contain_pulpcore__plugin('migration')
+            .that_subscribes_to('Class[Pulpcore::Install]')
+          is_expected.to contain_pulpcore__plugin('migration')
```
Comment (ekohl) [inline spec/classes/plugin_migration_spec.rb]: This shouldn't be needed. You can chain them

**Comment by ekohl** [comment]
>   Notice: /Stage[main]/Pulpcore::Database/Pulpcore::Admin[migrate --noinput]/Exec[python3-django-admin migrate --noinput]/returns: pkg_resources.DistributionNotFound: The 'importlib_metadata' distribution was not found and is required by jsonschema

@evgeni this suggests a packaging issue

**Comment by evgeni** [comment]
I agree, will look into it.

File: spec/classes/plugin_migration_spec.rb (hunk 6 above)
Comment (wbclark) [reply spec/classes/plugin_migration_spec.rb]: @ekohl I noticed that some were not getting executed when I did this. However removing the chaining did not resolve it. Is something excluding redundant tests?

**Comment by evgeni** [comment]
@ekohl packaging fix in https://github.com/theforeman/foreman-packaging/pull/4677

File: spec/classes/plugin_migration_spec.rb (hunk 6 above)
Comment (wbclark) [reply spec/classes/plugin_migration_spec.rb]: @ekohl after trying a few different approaches, I believe that the chaining works however the full output is suppressed. Therefore I have reverted to the chained approach.

**Comment by wbclark** [comment]
Will update this further with additional options from https://github.com/pulp/pulp-2to3-migration#configuration

**Pull request merged**